kquery/KQuery.g4

```diff
@@ -0,0 +1,357 @@
+    | LeftParen bitwiseExpr widthOrSizeExpr leftExpr rightExpr RightParen                   #BitwExprWidth
+    | LeftParen comparisonExpr widthOrSizeExpr leftExpr rightExpr RightParen                #CompExprWidth
+    | LeftParen comparisonExpr leftExpr rightExpr RightParen                                #CompExpr
+    | LeftParen concatExpr widthOrSizeExpr leftExpr rightExpr RightParen                    #ConcatExprWidth
+    | LeftParen concatExpr leftExpr rightExpr RightParen                                    #ConcatExprNW   
+    | LeftParen arrExtractExpr widthOrSizeExpr number expression RightParen                 #ArrExtractExprWidth
+    | LeftParen bitExtractExpr widthOrSizeExpr expression RightParen                        #BitExtractExprWidth
+    | LeftParen genericBitRead widthOrSizeExpr expression version RightParen                #ReadExpresssionVersioned
+    | LeftParen selectExpr widthOrSizeExpr leftExpr rightExpr expression RightParen         #SelectExprWidth
+    | LeftParen exprNegation widthOrSizeExpr expression RightParen                          #NegationExprWidth
+    | LeftParen exprNegation expression RightParen                                          #NegetionExpr
+    | LeftParen genericBitRead widthOrSizeExpr expression RightParen                        #ReadExpr
+    | version                                                                               #VersionExpr
+    | number                                                                                #Singleton
+    ;
+
+genericBitRead
+    : READ          
+    | READLSB
+    | READMSB
+    ;
+
+bitExtractExpr
+    : ZEXT
+    | SEXT
+    ;
+    
+version
+    : varName                                               #VersionVariableName
+    | namedConstant Colon expression                        #VersionedNamedAbbreviation
+    | LeftBracket updateList RightBracket ATR version       #UpdationList
+    | LeftBracket RightBracket ATR version                  #NoUpdateList
+    ;
+    
+notExpr
+    : NOT
+    ;
+    
+concatExpr
+    : CONCAT
+    ;
+
+exprNegation
+    : NEGETION
+    ;
+
+selectExpr
+    : SELECT
+    ;
+
+arrExtractExpr
+    : EXTRACT
+    ;
+    
+varName
+    : Identifier
+    ;
+    
+leftExpr 
+    : expression
+    ; 
+
+rightExpr
+    : expression
+    ;
+    
+namedConstant
+    : Identifier
+    ;
+    
+updateList 
+    : expression Equal expression COMMA updateList
+    | expression Equal expression
+    ;
+
+bitwiseExpr 
+    : BITWISEAND 
+    | BITWISEOR 
+    | BITWISEXOR 
+    | SHL 
+    | LSHR 
+    | ASHR
+    ;
+    
+comparisonExpr 
+    : EQ 
+    | NEQ 
+    | ULT
+    | UGT
+    | ULE 
+    | UGE 
+    | SLT 
+    | SLE 
+    | SGT 
+    | SGE
+    ;
+    
+arithmeticExpr 
+    : ADD
+    | SUB 
+    | MUL 
+    | UDIV 
+    | UREM 
+    | SDIV 
+    | SREM
+    ;
+    
+domain : widthOrSizeExpr ;
+rangeLimit : widthOrSizeExpr ;
+arrName : Identifier ;
+
+numberList
+    : number
+    | number numberList
```

Comment:
You really should write this rule in EBNF. `numberList: number+;`. This isn't just cosmetic, it's essential for a fast implementation--which you don't have. It's really, really slow.

Reply:
Yeah I had this rule `numberList: (number)+` earlier. I shall put that back. The recursive structure does make it slow for large list of numbers.